Support DCN parallelism

jonb377 opened the pull request:
- Add `spmd_dcn_parallelism` to define dcn parallelism.
- Reuse a single mesh across all sharding
- Shard parameters in `__init__` to avoid host-side OOM

JackCaoG commented on examples/pytorch/language-modeling/run_clm.py:
```diff
@@ -543,35 +551,51 @@ def get_mesh(ici_mesh_shape, dcn_mesh_shape=None):
+          # Place DCN parallelism on an independent axis
+          dcn_mesh_shape = (model_args.spmd_dcn_parallelism,) + (1,) * len(param.shape)
+          ici_mesh_shape = (1,) + tuple(mesh_shape)
+          mesh = get_mesh(ici_mesh_shape, dcn_mesh_shape)
```
what will get_mesh return here?

jonb377 replied:
Ah, I need to cut the FSDP dim by `dcn_parallelism`. The result should be `(dcn, n_dev // dcn, 1)` if max_dim = 0

JackCaoG replied:
hmm, what does this `get_mesh` function and why do we need to pass both `ici` and `dcn` mesh shapes?

jonb377 replied:
`get_mesh` allows us to switch between the iota mesh and HybridMesh. HybridMesh requires both `ici` and `dcn` shapes as input, so we need to keep them separate here as well

jonb377 replied:
The return is either `xs.Mesh` or `xs.HybridMesh` depending on the `spmd_iota_mesh` flag

JackCaoG replied:
ah nice

jonb377 commented:
Converting to draft for now, I need to update activation sharding as well

alanwaketan commented:
Okay, give me one moment before my PR is ready. Can you land #26? And I'm going to rebase on it.

jonb377 commented:
> Okay, give me one moment before my PR is ready. Can you land #26? And I'm going to rebase on it.

Done, thanks @alanwaketan

jonb377 committed:
Support DCN parallelism

defer init compatibility

Re-enable eval loop

Shard logits before loss calculation

Combind DCN and batch axis for activation sharding

Shard when initializing the model

Use the same mesh for input sharding

Fix input partition spec

Remove empty weight initialization

alanwaketan reviewed (approved):
LGTM, but let's us hold this until we finalize all the changes needed for Next.

alanwaketan commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -195,39 +195,34 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
         self.act_fn = ACT2FN[config.hidden_act]
 
-        init_spmd(self, config)
+        if self.spmd_fsdp_sharding:
```
Are we dropping tensor sharding?

alanwaketan replied:
I feel like we should drop it as it's not useful anyway and then it's not compatible with the 2D shading scheme. Let me do a follow up to drop it later.

alanwaketan commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -195,39 +195,34 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
-        init_spmd(self, config)
+        if self.spmd_fsdp_sharding:
+            gate_proj_spec = up_proj_spec = down_proj_spec = ('data', None)
+        else:
```
Can we comment that this is for 2D sharding?

jonb377 replied:
Sure, I'll add some more comments throughout

alanwaketan commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -312,13 +309,22 @@ def __init__(self, config: LlamaConfig):
         self._init_rope()
 
-        init_spmd(self, config)
+        if self.spmd_fsdp_sharding:
```
Same comments as above.

alanwaketan reviewed (approved):
LGTM.

jonb377 committed:
Add some comments

jonb377 committed:
Remove defer-init flag
